File: zookeeper/server/quorum/FollowerZooKeeperServer.cpp
/*
 * FollowerZooKeeperServer.cpp
 *
 *  Created on: 2017-11-22

 */

#include "./Follower.hh"
#include "./QuorumPeer.hh"
#include "./FollowerZooKeeperServer.hh"
#include "./FollowerRequestProcessor.hh"
#include "./SendAckRequestProcessor.hh"

namespace efc {
namespace ezk {

sp<ELogger> FollowerZooKeeperServer::LOG = ELoggerManager::getLogger("FollowerZooKeeperServer");

FollowerZooKeeperServer::FollowerZooKeeperServer(sp<FileTxnSnapLog> logFactory,QuorumPeer* self,
		sp<ZKDatabase> zkDb) THROWS(EIOException) :
		LearnerZooKeeperServer(logFactory, self->tickTime, self->minSessionTimeout,
			self->maxSessionTimeout, zkDb, self) {
	//
}

sp<Follower> FollowerZooKeeperServer::getFollower(){
	return self->follower;
}

void FollowerZooKeeperServer::setupRequestProcessors() {
	sp<RequestProcessor> finalProcessor = new FinalRequestProcessor(shared_from_this());
	commitProcessor = new CommitProcessor(finalProcessor,
			ELLong::toString(getServerId()), true,
			getZooKeeperServerListener());
	EThread::setDaemon(commitProcessor, true); //!!!
	commitProcessor->start();
	sp<FollowerRequestProcessor> frp = new FollowerRequestProcessor(dynamic_pointer_cast<FollowerZooKeeperServer>(shared_from_this()), commitProcessor);
	firstProcessor = frp;
	EThread::setDaemon(frp, true); //!!!
	frp->start();
	syncProcessor = new SyncRequestProcessor(shared_from_this(),
			new SendAckRequestProcessor(dynamic_pointer_cast<Learner>(getFollower())));
	EThread::setDaemon(syncProcessor, true); //!!!
	syncProcessor->start();
}

sp<Learner> FollowerZooKeeperServer::getLearner() {
	return getFollower();
}

void FollowerZooKeeperServer::logRequest(sp<TxnHeader> hdr, sp<ERecord> txn) {
	sp<Request> request = new Request(null, hdr->getClientId(), hdr->getCxid(),
			hdr->getType(), null, null);
	request->hdr = hdr;
	request->txn = txn;
	request->zxid = hdr->getZxid();
	if ((request->zxid & 0xffffffffL) != 0) {
		pendingTxns.add(request);
	}
	syncProcessor->processRequest(request);
}

void FollowerZooKeeperServer::commit(llong zxid) {
	if (pendingTxns.size() == 0) {
		LOG->warn("Committing " + ELLong::toHexString(zxid)
				+ " without seeing txn");
		return;
	}
	llong firstElementZxid = pendingTxns.element()->zxid;
	if (firstElementZxid != zxid) {
		LOG->error("Committing zxid 0x" + ELLong::toHexString(zxid)
				+ " but next pending txn 0x"
				+ ELLong::toHexString(firstElementZxid));
		ESystem::exit(12);
	}
	sp<Request> request = pendingTxns.remove();
	commitProcessor->commit(request);
}

void FollowerZooKeeperServer::sync(){
	SYNCHRONIZED(this) {
		if(pendingSyncs.size() ==0){
			LOG->warn("Not expecting a sync.");
			return;
		}

		sp<Request> r = pendingSyncs.remove();
		commitProcessor->commit(r);
	}}
}

int FollowerZooKeeperServer::getGlobalOutstandingLimit() {
	return LearnerZooKeeperServer::getGlobalOutstandingLimit() / (self->getQuorumSize() - 1);
}

void FollowerZooKeeperServer::shutdown() {
	LOG->info("Shutting down");
	try {
		LearnerZooKeeperServer::shutdown();
	} catch (EException& e) {
		LOG->warn("Ignoring unexpected exception during shutdown", e);
	}
	try {
		if (syncProcessor != null) {
			syncProcessor->shutdown();
		}
	} catch (EException& e) {
		LOG->warn("Ignoring unexpected exception in syncprocessor shutdown",
				e);
	}
}

} /* namespace ezk */
} /* namespace efc */

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FollowerZooKeeperServer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.zookeeper.server.quorum;

import java.io.IOException;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.LinkedBlockingQueue;
import javax.management.JMException;
import org.apache.jute.Record;
import org.apache.zookeeper.jmx.MBeanRegistry;
import org.apache.zookeeper.metrics.MetricsContext;
import org.apache.zookeeper.server.ExitCode;
import org.apache.zookeeper.server.FinalRequestProcessor;
import org.apache.zookeeper.server.Request;
import org.apache.zookeeper.server.RequestProcessor;
import org.apache.zookeeper.server.ServerMetrics;
import org.apache.zookeeper.server.SyncRequestProcessor;
import org.apache.zookeeper.server.ZKDatabase;
import org.apache.zookeeper.server.persistence.FileTxnSnapLog;
import org.apache.zookeeper.txn.TxnDigest;
import org.apache.zookeeper.txn.TxnHeader;
import org.apache.zookeeper.util.ServiceUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Just like the standard ZooKeeperServer. We just replace the request
 * processors: FollowerRequestProcessor -&gt; CommitProcessor -&gt;
 * FinalRequestProcessor
 *
 * A SyncRequestProcessor is also spawned off to log proposals from the leader.
 */
public class FollowerZooKeeperServer extends LearnerZooKeeperServer {

    private static final Logger LOG = LoggerFactory.getLogger(FollowerZooKeeperServer.class);

    /*
     * Pending sync requests
     */ ConcurrentLinkedQueue<Request> pendingSyncs;

    /**
     * @throws IOException
     */
    FollowerZooKeeperServer(FileTxnSnapLog logFactory, QuorumPeer self, ZKDatabase zkDb) throws IOException {
        super(logFactory, self.tickTime, self.minSessionTimeout, self.maxSessionTimeout, self.clientPortListenBacklog, zkDb, self);
        this.pendingSyncs = new ConcurrentLinkedQueue<Request>();
    }

    public Follower getFollower() {
        return self.follower;
    }

    @Override
    protected void setupRequestProcessors() {
        RequestProcessor finalProcessor = new FinalRequestProcessor(this);
        commitProcessor = new CommitProcessor(finalProcessor, Long.toString(getServerId()), true, getZooKeeperServerListener());
        commitProcessor.start();
        firstProcessor = new FollowerRequestProcessor(this, commitProcessor);
        ((FollowerRequestProcessor) firstProcessor).start();
        syncProcessor = new SyncRequestProcessor(this, new SendAckRequestProcessor(getFollower()));
        syncProcessor.start();
    }

    LinkedBlockingQueue<Request> pendingTxns = new LinkedBlockingQueue<Request>();

    public void logRequest(TxnHeader hdr, Record txn, TxnDigest digest) {
        Request request = new Request(hdr.getClientId(), hdr.getCxid(), hdr.getType(), hdr, txn, hdr.getZxid());
        request.setTxnDigest(digest);
        if ((request.zxid & 0xffffffffL) != 0) {
            pendingTxns.add(request);
        }
        syncProcessor.processRequest(request);
    }

    /**
     * When a COMMIT message is received, eventually this method is called,
     * which matches up the zxid from the COMMIT with (hopefully) the head of
     * the pendingTxns queue and hands it to the commitProcessor to commit.
     * @param zxid - must correspond to the head of pendingTxns if it exists
     */
    public void commit(long zxid) {
        if (pendingTxns.size() == 0) {
            LOG.warn("Committing " + Long.toHexString(zxid) + " without seeing txn");
            return;
        }
        long firstElementZxid = pendingTxns.element().zxid;
        if (firstElementZxid != zxid) {
            LOG.error("Committing zxid 0x" + Long.toHexString(zxid)
                      + " but next pending txn 0x" + Long.toHexString(firstElementZxid));
            ServiceUtils.requestSystemExit(ExitCode.UNMATCHED_TXN_COMMIT.getValue());
        }
        Request request = pendingTxns.remove();
        request.logLatency(ServerMetrics.getMetrics().COMMIT_PROPAGATION_LATENCY);
        commitProcessor.commit(request);
    }

    public synchronized void sync() {
        if (pendingSyncs.size() == 0) {
            LOG.warn("Not expecting a sync.");
            return;
        }

        Request r = pendingSyncs.remove();
        if (r instanceof LearnerSyncRequest) {
            LearnerSyncRequest lsr = (LearnerSyncRequest) r;
            lsr.fh.queuePacket(new QuorumPacket(Leader.SYNC, 0, null, null));
        }
        commitProcessor.commit(r);
    }

    @Override
    public int getGlobalOutstandingLimit() {
        int divisor = self.getQuorumSize() > 2 ? self.getQuorumSize() - 1 : 1;
        int globalOutstandingLimit = super.getGlobalOutstandingLimit() / divisor;
        return globalOutstandingLimit;
    }

    @Override
    public String getState() {
        return "follower";
    }

    @Override
    public Learner getLearner() {
        return getFollower();
    }

    /**
     * Process a request received from external Learner through the LearnerMaster
     * These requests have already passed through validation and checks for
     * session upgrade and can be injected into the middle of the pipeline.
     *
     * @param request received from external Learner
     */
    void processObserverRequest(Request request) {
        ((FollowerRequestProcessor) firstProcessor).processRequest(request, false);
    }

    boolean registerJMX(LearnerHandlerBean handlerBean) {
        try {
            MBeanRegistry.getInstance().register(handlerBean, jmxServerBean);
            return true;
        } catch (JMException e) {
            LOG.warn("Could not register connection", e);
        }
        return false;
    }

    @Override
    protected void registerMetrics() {
        super.registerMetrics();

        MetricsContext rootContext = ServerMetrics.getMetrics().getMetricsProvider().getRootContext();

        rootContext.registerGauge("synced_observers", self::getSynced_observers_metric);

    }

    @Override
    protected void unregisterMetrics() {
        super.unregisterMetrics();

        MetricsContext rootContext = ServerMetrics.getMetrics().getMetricsProvider().getRootContext();
        rootContext.unregisterGauge("synced_observers");

    }

}
